dags/common/pull_ftp.py

```diff
@@ -36,38 +38,51 @@ def migrate_from_ftp(
 ):
     params = kwargs["params"]
     if "force_pull" in params and params["force_pull"]:
-        return _force_pull(sftp, repo, logger)
+        return _force_pull(sftp, repo, logger, **kwargs)
     elif (
         "filenames_pull" in params
         and params["filenames_pull"]["enabled"]
         and params["filenames_pull"]["force_from_ftp"]
     ):
-        return _filenames_pull(params["filenames_pull"], sftp, repo, logger)
-    return _differential_pull(sftp, repo, logger)
+        return _filenames_pull(params["filenames_pull"], sftp, repo, logger, **kwargs)
+    return _differential_pull(sftp, repo, logger, **kwargs)
 
 
 def reprocess_files(repo: IRepository, logger: PrintLogger, **kwargs):
     filenames_pull_params = kwargs["params"]["filenames_pull"]
     filenames = filenames_pull_params["filenames"]
+    excluded_folders = kwargs["params"]["excluded_dirs"]
+    files_not_from_excluded_folder = remove_excluded_folders(
+        filenames, excluded_folders
+    )
     logger.msg("Processing specified filenames.", filenames=filenames)
-    return _find_files_in_zip(filenames, repo)
+    return _find_files_in_zip(files_not_from_excluded_folder, repo)
 
 
-def _force_pull(sftp: SFTPService, repo: IRepository, logger: PrintLogger):
+def _force_pull(sftp: SFTPService, repo: IRepository, logger: PrintLogger, **kwargs):
     logger.msg("Force Pulling from SFTP.")
     filenames = sftp.list_files()
-    return migrate_files(filenames, sftp, repo, logger)
+    excluded_folders = kwargs["params"]["excluded_dirs"]
+    files_not_from_excluded_folder = remove_excluded_folders(
```

Comment:
should I move somewhere else? It repeats 3 times, but I am not sure where is a better place to do this filtering. All of these passes result in migrate_files, but I believe that migrate_files function should receive the final list of file names, and start migrating from sftp to s3 :
https://github.com/cern-sis/workflows/blob/2b82a0178554fbb2a86217f5d459a8670b83c4a0/dags/common/pull_ftp.py#L54-L59

https://github.com/cern-sis/workflows/blob/2b82a0178554fbb2a86217f5d459a8670b83c4a0/dags/common/pull_ftp.py#L79-L83

https://github.com/cern-sis/workflows/blob/2b82a0178554fbb2a86217f5d459a8670b83c4a0/dags/common/pull_ftp.py#L109-L112

Reply:
Why we need a function for that? can we just pass a list of the excluded folders to the download function?

Reply:
migrate_files is the download function, but I am not really sure, should we just pass folder there. Because before it, we have to collect files names, and then pass to migrate_files.And we have 4 different collections of files:  take just missing files, pull force all, pull force spec.files. or reprocess files
 Different collections have logs different logs with list of files, for example:
    logger.msg("Pulling specified filenames from SFTP", filenames=filenames)
    logger.msg("Pulling specified filenames from SFTP", filenames=filenames)

Or is is too much repetitive and just filter files in migrate_files function? But then we will loose detailed logs that I mentioned before

Reply:
The collection of files comes from sftp function list_files. So it the begging it returns all file paths, and then we filter the ones we need. This filtered list is passed to download function (called migrate_files)

Reply:
because we are getting all file paths from sftp, and filtering just after we got it. I moved filtering  to migrate_files func, where the files are downloaded

Reply:
what you are saying is about the existing method `list_files` right?

Reply:
yes

Reply:
if you check the [documentation](https://pysftp.readthedocs.io/en/release_0.2.9/pysftp.html#pysftp.Connection.walktree) one of the callbacks can be used to ignore directories, could you please explore this solution? If we do that we just need to pass the excluded dirs to `list_files` and nothing else.

Reply:
we cannot customize it like this, or we need to write our own walktree method:
https://stackoverflow.com/questions/68392997/pysftp-connection-walktree-fails-if-any-directory-dont-have-read-permission. 
But I can filter in store_files_name func.

Reply:
updated: we don't need excluded_dirs in sftp, we can just filter it list_files by passing the variable of excluded_dirs there.